src/host.cpp

```diff
@@ -284,101 +256,98 @@ namespace rhost {
+            respond_to_message(msg, static_cast<double>(id));
+        }
 
-            if (msg.blob.size() != msg.blob_count) {
-                fatal_error("Incomplete blob creation message %s.", msg.id.c_str());
-            }
+        blob_id create_blob(blob&& blob) {
+            std::lock_guard<std::mutex> lock(blobs_mutex);
+            blob_id id = ++next_blob_id;
 
-            {
-                std::lock_guard<std::mutex> lock(received_blobs_mutex);
-                int blob_id = ++next_blob_id;
-                received_blobs[blob_id] = msg.blob;
-                respond_to_message(msg, static_cast<double>(blob_id));
+            // Check that it never overflows double mantissa, and provide immediate diagnostics if it happens.
+            if (id != blob_id(double(id))) {
+                fatal_error("Blob ID overflow");
             }
-        }
 
-        int create_blob(const rhost::util::blob& blob) {
-            std::lock_guard<std::mutex> lock(received_blobs_mutex);
-            int blob_id = ++next_blob_id;
-            received_blobs[blob_id] = blob;
-            return blob_id;
+            blobs[id] = std::move(blob);
+            return id;
         }
 
-        const blob_slice get_blob_as_single_slice(int blob_id) {
-            std::map<int, rhost::util::blob>::iterator res = received_blobs.find(blob_id);
-            blob_slice data;
-
-            if (res == received_blobs.end()) {
-                return data;
-            }
-
-            size_t totalSize = 0;
-            for (blob_slice& slice : res->second) {
-                totalSize += slice.size();
-            }
+        bool get_blob(blob_id id, blobs::blob& blob) {
+            std::lock_guard<std::mutex> lock(blobs_mutex);
+            auto& it = blobs.find(id);
 
-            data.reserve(totalSize);
-            for (blob_slice& slice : res->second) {
-                data.insert(data.end(), slice.begin(), slice.end());
+            if (it == blobs.end()) {
+                return false;
             }
 
-            return data;
+            blob = it->second;
+            return true;
         }
 
-        void get_blobs(const message& msg) {
-            assert(msg.name == "?GetBlob");
+        void get_blob(const message& msg) {
+            assert(!strcmp(msg.name(), "?GetBlob"));
 
-            blob blob;
-            picojson::array blob_ids;
-            for (int i = 0; i < msg.args.size(); ++i) {
-                int blob_id = static_cast<int>(msg.args[i].get<double>());
-                blob_slice slice = get_blob_as_single_slice(blob_id);
-                if (slice.size() > 0) {
-                    blob.push_back(slice);
-                    rhost::util::append(blob_ids, static_cast<double>(blob_id));
-                }
+            auto json = msg.json();
+            if (!json[0].is<double>()) {
+                fatal_error("GetBlob: non-numeric blob ID");
+            }
+            auto id = static_cast<blob_id>(json[0].get<double>());
```

Comment:
I can use `double` everywhere, but I really don't want to, because it's not the true value domain (then everything would have to check for non-integer doubles and raise errors, or just ignore and round with potentially strange behavior). Double is really used as a 52-bit int for transport purposes here, and is due to transport limitations. Long-term, I'm hoping to get rid of the JSON issue by moving to MsgPack (similar structure, but binary serialization, and natively supports blobs, 64-bit ints etc). And there might be a better way to represent this in R as well, e.g. as a pair of 32-bit ints.